### PR TITLE
POC: Add GraphQL API for query service

### PR DESCRIPTION
**Pull request by pavolloffay**
Updates https://github.com/jaegertracing/jaeger/issues/169

Please do not nit-pick this is a POC with rather ugly code. I just want to demonstrate how GraphQL API might work with the query service.

#### Goals
* Documented and stable query API
* Return OTLP JSON proto model - https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/collector/trace/v1/trace_service.proto#L41 
  * Use "standard"/well defined model
  * The returned data can be directly reingested 
 
##### https://github.com/99designs/gqlgen/

* GraphQL SDL first approach 
* Tool `gqlgen` generates handlers (and optionally model) from schema file

Issues:
* In order to be compatible with OTLP trace JSON `JSONPb` marshaller has to be used. GraphQL framework allows to use  custom marshallers, however this works only well for scalar types. If the marshaller is used for objects then selecting properties/fields in the query does not work (e.g. `"query": "query {trace(traceId: \"todo\") { resourceSpans{instrumentationLibrarySpans{spans{name}}} }}"}` to select the span name). There is a manual and a quite ugly workaround (I will add a pointer to the code), also note that GraphQL query has to specify at least one object propery. 
* The GraphQL response wraps the returned JSON into the following object. Hence the response cannot be directly processed and ingested as OTPL JSON traces.

```json
{
  "data": {
    "<queryName>/getTrace": {
        "resourceSpans": {}
    }
}
```

##### Other notes

Example curl request
```
curl -ivX POST -H 'Content-Type: application/json' -d '{"query": "query {trace(traceId: \"<traceId>\") { resourceSpans{instrumentationLibrarySpans{spans{name}}} }}"}' http://localhost:8080/query
```

Example GraphQL queries in the GraphQL editor `http://localhost:8080/`
```graphql
query getTrace{ trace(traceId: "foo") { resourceSpans{instrumentationLibrarySpans{ instrumentationLibrary{name} spans{traceId}}}  } }

query findTraces {
  traces(service: "payments", startMicros: 10, endMicros:150) {
    name
    traceId
  }
}
```

### CODE REVIEW

File: cmd/query/app/graphql/graph/schema.resolvers.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		TraceId: []byte{0, 1, 2, 3, 4},
+		Name:    "hello",
+	}
+
+	// Remove fields that hasn't been requested
+	// this is a workaround because:
+	// * a custom marshaller is used for OTLP Span
+	//   to ensure compatibility with OTLP JSON (e.g. standard JSON marshalling cannot be used for protos)
+	// * a custom marshaller works well only with scalar types
+	if !names["resourceSpans.instrumentationLibrarySpans.spans.traceId"] {
```
Comment (pavolloffay) [inline cmd/query/app/graphql/graph/schema.resolvers.go:51]: Here is the workaround to make selecting object properties/fields work. This is only needed when custom object marshaller is used. The custom marshaller `JSONPb` for `ResourceSpans` has to be used to be compatible with OTLP trace JSON.

**Comment by yurishkuro** [comment]
btw, do we know how https://github.com/dgraph-io/dgraph is going graphql with Go & gRPC?

**Comment by pavolloffay** [comment]
by looking at dependencies it might be using https://github.com/graph-gophers/graphql-go

**Comment by pavolloffay** [comment]
Closing in favour of https://github.com/jaegertracing/jaeger/pull/3086

**Pull request closed without merging**